lib/dao.js

```diff
@@ -624,7 +624,193 @@ DataAccessObject.upsert = function(data, options, cb) {
+            callConnector();
+          } else {
+            inst.isValid(function(valid) {
+              if (!valid) {
+                if (doValidate) {
+                  return cb(new ValidationError(inst), inst);
+                } else {
+                  console.warn('Ignoring validation errors in updateOrCreate():');
+                  console.warn('  %s', new ValidationError(inst).message);
+                }
+              }
+              callConnector();
+            }, update, options);
+          }
+
+          function callConnector() {
+            update = removeUndefined(update);
+            context = {
+              Model: Model,
+              where: ctx.where,
+              data: update,
+              currentInstance: inst,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('persist', context, function(err) {
+              if (err) return done(err);
+              connector.upsertWithWhere(modelName, where, update, options, done);
+            });
+          }
+          function done(err, data, info) {
+            if (err) return cb(err);
+            var context = {
+              Model: Model,
+              data: data,
+              isNewInstance: info && info.isNewInstance,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('loaded', context, function(err) {
+              if (err) return cb(err);
+              var obj;
+              if (data && !(data instanceof Model)) {
+                inst._initProperties(data, { persisted: true });
+                obj = inst;
+              } else {
+                obj = data;
+              }
 
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            });
+          }
+        });
+      } else {
+        var element = Object.keys(where)[0];
+        var filter = { where: where };
+        if (!Object.keys(data).length) {
+          var err = new Error('data object cannot be empty!');
+          err.statusCode = 400;
+          process.nextTick(function() { cb(err); });
+        }
+
+        if (!data.hasOwnProperty(element)) {
+          data[element] = where[element];
+        }
+        var primaryKey = idName(Model);
+        self.find(filter, function(err, instances) {
+          if (err) cb(err);
+          else {
+            var modelsLength = instances.length;
+            if (modelsLength === 0 && data.hasOwnProperty(primaryKey))  {
+              return self.create(data, options, cb);
+            } else if (modelsLength == 0 && data.hasOwnProperty(primaryKey) == false) {
```

Comment:
I have mentioned [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001#discussion_r70986246) and [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001#discussion_r71027682)

`id` can be auto generated `id` in many cases and not provided in `data` object; you should take care of this scenario.